src/main/flight/imu.c

```diff
@@ -215,51 +213,79 @@ static float invSqrt(float x)
+
+        // Compute IMU / craft heading vector and magnitude in body X axis
+        // Important:  vector magnitude is reduced as pitch increases
+        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};
+        const float headingMagnitude = vector2Mag(&heading_ef);
+
+        // cross (product) = heading * cogHeading * sin error angle, ie, reflects error angle of quad X axis vs GPS groundspeed
+        // cross value sign depends on the rotation change from input to output vectors by right hand rule
+        // operand order is such that rotation is in direction of zero error
+        // abs value of cross is zero when parallel, max of 1.0 at 90 degree error
+        // cross value determines ez_ef when error is less than 90 degrees
+        // a zero cross value at 180 degree error would prevent error correction
         const float cross = vector2Cross(&heading_ef, &cog_ef);
-        // dot product, 1 * |heading| * cos(angle)
+
+        // dot product = heading * cogHeading * cos(angle)
+        // max when aligned, zero at 90 degrees of vector difference, negative for any error > 90 degrees
         const float dot = vector2Dot(&heading_ef, &cog_ef);
-        // use cross product / sin(angle) when error < 90deg (cos > 0),
-        //   |heading| if error is larger (cos < 0)
-        const float heading_mag = vector2Mag(&heading_ef);
-        float ez_ef = (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) * heading_mag;
-#if THRUST_COG
-        // increase gain for small tilt (just heuristic; sqrt is cheap on F4+)
-        ez_ef /= sqrtf(heading_mag);
-#endif
-        ez_ef *= cogYawGain;          // apply gain parameter
+
+
+        // ez_ef is the error value used to correct the heading via an IIR like smoothing function
+        // for error angles less than 90 degrees (dot > 0), use cross value to compute ez_ef
+        // above 90 degrees error, use heading vector magnitude, but give it sign from cross product
+        // heading vector magnitude is 1.0 when flat, and is attenuated when pitch angle is large
+        // when error exceeds 90 degrees, and pitch angle is high, a transient step may occur in ez_ef
+        float ez_ef = (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) * headingMagnitude;
+
+        // avoid using GPS (reduce ez_ef) during and soon after roll or yaw inputs
+        // because the GPS heading will tend to lag the true heading
+        const float deflectionRPMaxInv = 1.0f - fmaxf(getRcDeflectionAbs(FD_ROLL), getRcDeflectionAbs(FD_YAW));
+        float deflectionRPMaxInvPwr5 = power5(deflectionRPMaxInv);
+        // negative peak detector with 1s time constant 1.0s
+        static float prev = 0.0f;
+
+        const float k = 0.5f * dt; // k = 0.005 at 100Hz, dt in seconds, 2.0s time constant
+        float smoothedStickMovement = prev + k * (deflectionRPMaxInvPwr5 - prev);
+        prev = fminf(smoothedStickMovement, deflectionRPMaxInvPwr5);
+
```

Comment:
This must be outside of IMU code

Reply:
see above

Reply:
It is important to ensure that the IMU estimate of heading is accurate.  We already use speed to inhibit ez_ef from responding to garbage when the quad moves slowly, and to increase ez_ef when it moves faster.  These things are required to ensure IMU heading estimate accuracy in normal flight.